handler/src/main/java/io/netty/handler/ssl/ResumptionController.java

```diff
@@ -0,0 +1,170 @@
+                        "Resumable trust manager must be a subclass of X509ExtendedTrustManager");
+            }
+            if (!resumableTm.compareAndSet(null, (ResumableX509ExtendedTrustManager) tm)) {
+                throw new IllegalStateException("Only one TrustManager can be configured for resumed sessions");
+            }
+            return new X509ExtendedWrapTrustManager((X509ExtendedTrustManager) tm);
+        } else if (logResumptionFailure) {
+            logger.debug("Not wrapping trust manager {} of type {}", tm, tm != null ? tm.getClass() : null);
+        }
+        return tm;
+    }
+
+    public void remove(SSLEngine engine) {
+        if (resumableTm.get() != null) {
+            confirmedValidations.remove(unwrapEngine(engine));
+        }
+    }
+
+    public boolean validateResumeIfNeeded(SSLEngine engine)
+            throws CertificateException, SSLPeerUnverifiedException {
+        ResumableX509ExtendedTrustManager tm;
+        boolean valid = engine.getSession().isValid();
+        if ((tm = resumableTm.get()) != null && valid) {
+            Certificate[] peerCertificates = engine.getSession().getPeerCertificates();
+            engine = unwrapEngine(engine);
+            if (confirmedValidations.remove(engine) == null) {
+                // This is a resumed session.
+                if (isClient) {
+                    tm.resumeClientTrusted(chainOf(peerCertificates), engine);
+                } else {
+                    tm.resumeServerTrusted(chainOf(peerCertificates), engine);
+                }
+                return true;
+            }
+        } else if (logResumptionFailure) {
+            logger.debug("Not resuming session, tm={}, valid={}", tm, valid); // TODO remove this logging
+        }
+        return false;
+    }
+
+    private static SSLEngine unwrapEngine(SSLEngine engine) {
+        if (engine instanceof JdkSslEngine) {
+            return ((JdkSslEngine) engine).getWrappedEngine();
+        }
+        return engine;
+    }
+
+    private static X509Certificate[] chainOf(Certificate[] peerCertificates) {
+        if (peerCertificates instanceof X509Certificate[]) {
+            //noinspection SuspiciousArrayCast
+            return (X509Certificate[]) peerCertificates;
+        }
+        X509Certificate[] chain = new X509Certificate[peerCertificates.length];
+        for (int i = 0; i < peerCertificates.length; i++) {
+            chain[i] = (X509Certificate) peerCertificates[i];
+        }
+        return chain;
+    }
+
+    @SuppressJava6Requirement(reason = "Guarded by version check")
+    private class X509ExtendedWrapTrustManager extends X509ExtendedTrustManager {
+        private final X509ExtendedTrustManager trustManager;
+
+        X509ExtendedWrapTrustManager(X509ExtendedTrustManager trustManager) {
+            this.trustManager = ObjectUtil.checkNotNull(trustManager, "trustManager");
+        }
+
+        private void unsupported() throws CertificateException {
```

Comment:
this could be static
```suggestion
        private static void unsupported() throws CertificateException {
```